references/classification/train.py

```diff
@@ -210,16 +215,22 @@ def main(args):
 
     collate_fn = None
     num_classes = len(dataset.classes)
-    mixup_transforms = []
+    mixup_or_cutmix = []
     if args.mixup_alpha > 0.0:
-        mixup_transforms.append(transforms.RandomMixup(num_classes, p=1.0, alpha=args.mixup_alpha))
+        mixup_or_cutmix.append(transforms.RandomMixup(alpha=args.mixup_alpha, p=1.0))
     if args.cutmix_alpha > 0.0:
-        mixup_transforms.append(transforms.RandomCutmix(num_classes, p=1.0, alpha=args.cutmix_alpha))
-    if mixup_transforms:
-        mixupcutmix = torchvision.transforms.RandomChoice(mixup_transforms)
+        mixup_or_cutmix.append(transforms.RandomCutmix(alpha=args.cutmix_alpha, p=1.0))
+    if mixup_or_cutmix:
+        batch_transform = transforms.Compose(
+            [
+                transforms.LabelToOneHot(num_categories=num_classes),
```

Comment:
I removed from this line the `WrapIntoFeatures()` call and now get:
```
TypeError: RandomCutmix() is only defined for tensor images and one-hot labels.
```

I expected that the `Label` is already annotated because the dataset is defined with:
```python
target_transform=lambda target: features.Label(target),
```

I'll check it out further on Monday.

Reply:
Yeah, that is a known issue. The `default_collate` internally uses [`torch.stack`](https://github.com/pytorch/pytorch/blob/9c036aa112b0a8fd9afb824d1fda058e2b66ba1d/torch/utils/data/_utils/collate.py#L143). Since we only automatically wrap on `.to()` and `.clone()` this will drop the feature type.

Three options here:

1. Allow `torch.stack` to retain the feature type. I would not go that route, since we cannot know in general if the result of a stack is still the feature we started out with. We at least need more elaborate checks that make sure this is happening.
2. Fix `default_collate` to handle our feature types. It is a possible solution, but won't happen, because the feature types are `torchvision` only and `default_collate` is a general solution.
3. Provide a custom collate like `vision_collate`. This would mean that depending on the taks, users can no longer go with the default value in the `DataLoader`, but judging from our references, this seems to be somewhat common. Plus, this `vision_collate` could also include more functionality that we need:
    1. Don't `torch.stack` images if we find `BoundingBox`'es.
    2. Support for `None` that some of our datasets already return (#5233) and will be heavily utilized by the new datasets for missing data in a configuration.

My vote is out for 3. To avoid rewriting `default_collate` from scratch in `torchvision` and thus risking it going out of sync with the core version, I made a push to make `default_collate` more extensible in [this gist](https://gist.github.com/pmeier/da684e84283113ed6e427d6519d21efe). @ejguan wanted to handle this from the `torchdata` side. Any progress there?

Reply:
It's a hard problem. I agree option 2 is not possible. What I don't like about option 3 is that if TorchVision has it's own collate, it will miss out on potential improvements from Core/Data. I also wouldn't put it inside the library but rather in the examples.

Option 1 sounds promising only if we feel we can do all the necessary checks to avoid gotchas (+ think of whether this is a performance bottleneck). If that's the case, we could offer implementations that whitelist the `stack` similary to `to()` and `clone()`. But if checking all the TV meta-data + Tensor meta-data is too slow, then we might leave it to the user to decide how to handle it efficiently on a case by case basis (as you have it here). I'm also open exploring other ideas to rewrap the types, though nothing elegant pops to my mind right now.

Reply:
Talked to Vitaly about it a while ago and we thought adding argument to `default_collate` might not be the ideal option because users can always do it in a reversed way by providing custom collate function and call `default_collate` if needed.

Would it be possible if we let `default_collate` to do collate by checking if the object is attached with a specific method (tentative: `collate_fn`)?
```py
class CustomObject:
   @staticmethod
    def collate_fn(batch):
        return ...

def default_collate(batch:
    elem = batch[0]
    ...
    if hasattr(elem, "collate_fn"):
        return elem.collate_fn(batch)
```
This would provide the [same mechanism](https://github.com/pytorch/pytorch/blob/6a04df3ac85714b33dcb2af20d0eeea96d131d75/torch/utils/data/_utils/pin_memory.py#L72-L73) as `pin_memory`

Reply:
> users can always do it in a reversed way by providing custom collate function and call `default_collate` if needed.

That is not true. `default_collate` does two things:

1. Handle containers such as lists, tuples, dictionaries, and so on, by recursively calling itself.
2. Handle "primitive batches", by `torch.stack`'ing tensors, putting integers and floats in a list and so on.

What we need is to modify 2. without touching 1. Since these two parts are inseparable at the moment, we cannot call `default_collate` to do 1. for us, because it would do the wrong "primitive batch" handling. Thus, we have to rewrite the functionality of 1. from scratch and can only call `default_collate` for the "primitive batches" that we don't want to handle differently.

> we thought adding argument to `default_collate` might not be the ideal option

My proposal is not advocating for that. I introduced a `collate` function that only does 1. from above and is flexible on 2. The old `default_collate` will just `functools.partial` of from it and thus not need / expose any extra parameters.

Reply:
As for adding a `collate_fn` to our feature type, I'm not really a fan. It feels a little weird to have this logic on the object, but I guess it is easier than to allow `torch.stack` to retain the feature type (option 1. from above).

Still, it can't be `@staticmethod` due to JIT, right @datumbox? But in @ejguan's example above, I don't see a reason why it needs to be, other than it makes little sense to have this as an instance method from the theoretical perspective.

Reply:
Ohhh you are right. So, those `Feature`s are primitive objects? I thought it's going to be a collection object that needs to handle dict or list specially at the outer level.

> I introduced a `collate` function that only does 1

I agree on adding this `collate` function to handle all collection objects before the primitive objects. But, in terms of how to register collate function for a certain primitive type of objects, I still don't agree on suppressing `TypeError`. IMHO, It's error-prone. Like [here](https://github.com/pytorch/pytorch/blob/f64857189d514a662dac09ca6421e1e95e87e843/torch/utils/data/_utils/collate.py#L148-L149), we do want to raise `TypeError` that will be silently suppressed by the `with contextlib.suppress(TypeError):`.

So, I think prefer the following two options:
1. Add this `collate` function with  an argument of `Dict[Type, collate_fn]` to handle each object based on the type
    - First check if `type(elem)` in the dict
    - Second go through each key in dict and check `isinstance` for derived class
2. Keeping the existing `default_collate`, but before any collate on primitive objects, we try to use `collate_fn` attached to the class.

In terms of JIT, I don't know if any of the above option would work. To make JIT working, a jittable custom collate function might be the best option to you.

Reply:
> * Second go through each key in dict and check `isinstance` for derived class

Shouldn't it be an `OrderedDict` then? Otherwise if an input is a subtype of multiple types in the map, two runs may yield different results making the collation non-deterministic.

--- 

I would prefer 1. since otherwise an object would need to know how it wants to be collated, but doesn't have all the context. For example, for classification and segmentation tasks, we want to `torch.stack` images, but for detection tasks they should be kept in a list. We can figure this out from the full sample passed to the collation function, but he object itself can't.

Plus, we don't have to worry about JIT and that can be quite a pain.

If you send a PR, please ping me there so I can see if it actually is extensible enough for us. @datumbox can you confirm that you would also be ok with handling this in the collation function if we only need to provide custom behavior for our stuff, but otherwise can depend on the core functionality?

Reply:
> Shouldn't it be an `OrderedDict` then?

If the `type` exists in the dict, it should invoke the corresponding collate function. And, `dict` is actually ordered in Python. So, if an input is a subtype of multiple types, and the type of the input doesn't equal to any of key, it will follow the order.

Will open a PR today. It should be a simple change.

Reply:
> And, `dict` is actually ordered in Python.

Well, not really. Since Python 3.6, `dict`'s are insertion ordered. Meaning, if you create a dictionary like

```py
dct = {"a": 1, "b": 2}
```

you can be sure that `"a"` is always the first key. However, `dict`'s are not ordered

```py
assert {"a": 1, "b": 2} == {"b": 2, "a": 1}
```

Reply:
If you try iterate over those dicts, you will find it's ordered.
```py
d1 = {"a": 1, "b": 2}
for k in d1:
    print(k)  # a, b

d2 = {"b": 2, "a": 1}
for k in d2:
    print(k)  # b, a
```

Reply:
>  @datumbox can you confirm that you would also be ok with handling this in the collation function if we only need to provide custom behavior for our stuff, but otherwise can depend on the core functionality?

Just to confirm, is your proposal to handle custom behaviour on our reference scripts? If yes, then that sounds good. I wouldn't want to add this on main TorchVision but rather leave it to the users to handle those special bases based on their use-case/dataset. Does that make sense?